### PR TITLE
Rust Integration: Perform Max Flow in Rust instead of Python

### PR DESCRIPTION
**Pull request by ohjuny**
The overarching goal is to improve performance by using Rust instead of Python to perform intense computations, while maintaining the Python API.

We implement Python-Rust interop using [`pyo3`](https://github.com/PyO3/pyo3/tree/main). We use the [`pathfinding`](https://docs.rs/pathfinding/latest/pathfinding/) crate for the Rust-side implementation of the Edmonds-Karp max flow algorithm.

Through initial profiling, we have identified the call to `nx.maximum_flow` to be the overwhelming bottleneck (51-74% of total time spent). This PR replaces the calls to `nx.maximum_flow` with Rust-side bindings, which initial profiling shows a total speedup of 1.6-2.6x.

### Links
* [Profiling Data](https://docs.google.com/spreadsheets/d/1CS7ba075mFsgYYo8KnIve6Tm_-cpZev2Lxp0kNV8bpU/edit?gid=0#gid=0)
* Profiling Infrastructure repo: https://github.com/ohjuny/lowtime-exp

### Results
Based on profiling runs of 5 models on A40 and A100 GPUs, we found:
* 3.83~6.99x speed up on first call to Rust-side `max_flow`
* 1.89~3.56x speed up on seconds call to Rust-side `max_flow`
* Total runtime: 1.31~3.46x speedup

## Build and Run
Assume you have some project that is using lowtime, and you have a local clone of this lowtime branch.
```bash
$ ls
project/    lowtime/
```
Then run the following to set up a virtual environment with this version of lowtime.
```bash
$ cd project/
$ python -m venv .env        # create virtual environment
$ source .env/bin/activate   # activate virtual environment
$ cd ../lowtime/
$ pip install -e .           # (editable) install lowtime
$ cd ../project/
$ #work on project, lowtime is now installed
```
If you modify `lib.rs` and want those changes reflected:
```bash
$ pwd
lowtime
$ pip uninstall lowtime
$ pip install -e .
```

### `SparseCapacity` vs `DenseCapacity`
The Rust-side code uses the [`pathfinding::durected::edmonds_karp::edmonds_karp`](https://docs.rs/pathfinding/latest/pathfinding/directed/edmonds_karp/fn.edmonds_karp.html) implementation for max flow. The function takes in an `EK` type parameter, that determines whether to represent the graph using a [`SparseCapacity`](https://docs.rs/pathfinding/latest/pathfinding/directed/edmonds_karp/struct.SparseCapacity.html) (`BTree` of `Btree`s) or [`DenseCapacity`](https://docs.rs/pathfinding/latest/pathfinding/directed/edmonds_karp/struct.DenseCapacity.html) (adjaceny matrix). Due to the nature of computation graphs of parallelized training workloads, a `SparseCapacity` was more efficient. This was verified through profiling, which showed that `SparseCapacity` was up to 10x faster than `DenseCapacity`.

### `OrderedFloat<f64>` vs `f32` vs `i64`
The `edmonds_karp` implementation requires the `Zero, `Bounded`, `Signed`, `Ord`, `Copy` traits to be implemented by capacity values.
* Raw floats in Rust do not implement `Ord`, so they cannot be used; `OrderedFloat` does.
* `OrderedFloat<f32>` resulted in incorrect output due to lack of precision.

We considered an implementation where we use integers, and convert to/from floats by multiplying by 1e9.
* Using `i64` resulted in incorrect output. Furthermore, it resulted in a similar runtime (no significant gain).
* `u64` cannot be used as it does not implement `Signed`, which is needed to represent negative residual flows in `edmonds_karp`.

Therefore `OrderedFloat<f64>` was chosen to represent capacity.

### `pyo3` Data Transfer Reduction with "pimpl" idiom
The `_lowtime_rs::PhillipsDessouky` constructor returns the object itself, which is very large as it contains all the nodes, edges, and associated capacities. Due to a misconception that `pyo3` would have to create a "Python copy" of the Rust object, we experimented with using the pimpl idiom to implement a wrapper `PhillipsDessouky` class that contained a pointer to an internal `_PhillipsDessouky` class, which contained the actual data and implementations. However, `pyo3` seems to have a clever way of associating Rust objects with Python objects without creating a "Python copy". In fact, profiling the "pimpl version" resulted in slightly slower runtimes, probably due to the added indirection and use of the heap.

### Handling Rust-side Errors
Before this PR, the lowtime codebase performs exception handling on the call to`nx.maximum_flow`, since the `nx` library is designed to raise well-defined exceptions (eg. `nx.NetworkXUnbounded`). However, the Rust-side `pathfinding` library is does not contain well-defined errors; instead, the `edmonds_karp` function considers the following situations unrecoverable:
* `panic!`s if `source` and `sink` nodes are not found in `verticies`.
* `unreachable!` if there is no flow to cancel in its internal representation of node capacities. One way this is caused is by providing a negative capacity as input (this caused a very hard-to-debug error when some flows were very small negative values due to floating point imprecision).
* `assert!`s many invariants of the internal graph representation (`assert!` will `panic!` if the condition is false).

If the Rust-side library use well-defined errors, `pyo3` has integration for propagating those errors as Python exceptions. The "recommended way" when using a third-party crate that does not contain well-defined errors is to, if possible, modify the third-party crate itself to do this.

`pathfinding` chooses to use unrecoverable macros over well-defined errors. Assuming we do not intend on changing the `pathfinding` codebase itself, we have 2 options:
1. Use `catch_unwind`to catch `panic!`s in Rust. Once caught, we convert the error to a custom Error type that is propagated to Python as an exception, which can be caught and handled in Python.
2. Do not handle `panic!`s, letting the program crash gracefully. The `panic!` and `unreachable!` macros are intended as graceful exit points for unrecoverable states, so we should not try to treat it like a Python exception.

There is significant debate on this topic; for example, this discussion based on the [polars](https://github.com/PyO3/pyo3/issues/2783) library. In the `polars` discussion, the debate was closed with:
<blockquote>
A Rust panic really is intended to be treated as a (graceful) crash, not as a user-space error. It sounds from the polars thread that the maintainers there are willing to treat panics as bugs and fix them.

> one does not need to manually extend error catching conditions when using a cython-wrapped library, as one would have to do when including a pyo3-wrapped library.

The point is that a panic isn't a typical error condition (returning Result is). A Python user isn't expected to interact with panics, the choice to not extend from Exception is deliberate so that panics are not accidentally caught.

Will close this one, if there is significant pressure from the community we can revisit.
</blockquote>

We make the same decision here: we do not attempt to catch or handle `panic!`s and `unreachable!` crashes. If needed in the future, we can add invariant checks before the call to `edmonds_karp`.

### Profiling Set Up
Through the development process, we use profiling data to determine whether an idea was worth pursuing. Since there were many ideas that depended on each other, there was a need to build a profiling infrastructure fast (even if it is scrappy) before we could start experimenting with lowtime. As a result, the current profiling inftrastructure was built:
1. Measure times between intervals of interest (eg. the call to `find_min_cut`).
2. Log the time.
3. Run a separate script after lowtime has completed that parses the `job.log` for profiling-related logs and groups relevant intervals together.

While fast, this approach results in many profiling related logs in the lowtime codebase, and requires a separate script to parse the logs post-lowtime. An ideal solution would be integrated profiling infrastructure integrated into lowtime itself, that can be turned on/off through a command-line argument when running lowtime. However, this PR chooses the "scrappy" option because:
1. The need for profiling data was blocking progress on the "core" idea (Rust interop).
2. Building an integrated profiling infrastructure seems like an independent addition that could be its own PR.

We will need to remove the profiling logs at the end of this chain of commits (i.e. when we eventually merge `lowtime-rust` with `main`), but we keep them for now as we will need to use the existing profiling infrastructure for future commits in this chain.

### CODE REVIEW

File: lowtime/lowtime_rust/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            edges_raw: edges_raw,
+        })
+    }
+
+    fn max_flow(&self) -> PyResult<Vec<((u32, u32), f64)>> {
+        let mut profiling_start;
+        let mut profiling_end;
+        // let SCALING_FACTOR: f64 = 1000000000.0;
+
+        profiling_start = Instant::now();
```
Comment (jaywonchung) [inline lowtime/lowtime_rust/src/lib.rs]: You can just write `let profiling_start = Instant::now();` everywhere. A new `let` declaration will **shadow** the old variable of the same name. See [this example](https://doc.rust-lang.org/rust-by-example/variable_bindings/scope.html).

File: lowtime/lowtime_rust/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        // ohjun: do we actually need to return this? This is unncessary data transfer from Rust->Python
+        // Instead, why not have a type that stores a pointer to a PhillipsDessouky, and we can just
+        // send the outer type? This way, Python can use the object to call max_flow and things, but
+        // we can avoid returning all this data here?
+        Ok(PhillipsDessouky {
+            node_ids: node_ids,
+            source_node_id: source_node_id,
+            sink_node_id: sink_node_id,
+            edges_raw: edges_raw,
+        })
```
Comment (jaywonchung) [inline lowtime/lowtime_rust/src/lib.rs]:
```suggestion
        Ok(PhillipsDessouky {
            node_ids,
            source_node_id,
            sink_node_id,
            edges_raw,
        })
```

Constructor shorthand. See the [Rust book](https://doc.rust-lang.org/beta/book/ch05-01-defining-structs.html#using-the-field-init-shorthand).

Comment (jaywonchung) [reply lowtime/lowtime_rust/src/lib.rs]: I think these aren't returned to Python BTW. It's just constructing the object from the Rust side, and Python will get a pointer to the Rust-side object. Whenever it access attributes (and it probably cannot since you didn't define `#[getter]`s for these attributes), it will call copy out the Rust-side data and convert it into Python objects.

File: lowtime/solver.py [hunk 3]
```diff
@@ -465,20 +567,58 @@ def find_min_cut(self, dag: nx.DiGraph) -> tuple[set[int], set[int]]:
...
+            rust_dag = lowtime_rust.PhillipsDessouky(
+                nodes, source_node, sink_node, edges
+            )
+            profiling_data_transfer = time.time() - profiling_data_transfer
+            logger.info(
+                "PROFILING PhillipsDessouky::find_min_cut data transfer 2 time: %.10fs",
+                profiling_data_transfer,
             )
+
+            rust_flow_vec = rust_dag.max_flow()
```
Comment (jaywonchung) [inline lowtime/solver.py]: Did you actually already replace the second max flow with Rust? How much speed up did you see?

**Comment by jaywonchung** [comment]
Could you also quickly document how to build and run this?

File: lowtime/solver.py (hunk 3 above)
Comment (ohjuny) [reply lowtime/solver.py]: I was getting some errors when I replaced the second call... will look into it further.

Comment (ohjuny) [reply lowtime/solver.py]: Successfully replaced second max flow with Rust. Summarizing all profiling results here: https://docs.google.com/spreadsheets/d/1CS7ba075mFsgYYo8KnIve6Tm_-cpZev2Lxp0kNV8bpU/edit?gid=0#gid=0

(Note: only you and I have access to the spreadsheet for now, can make it viewable publicly if that's fine)

File: rust/.github/workflows/CI.yml
Comment (jaywonchung) [inline rust/.github/workflows/CI.yml:1]: You should probably put this `.github` directory outside for it to get picked up

File: rust/Cargo.lock
Comment (jaywonchung) [inline rust/Cargo.lock:1]: For libraries, it is conventional to add `Cargo.lock` to `.gitignore` and not commit it into version control.

**Comment by ohjuny** [comment]
Not 100% knowledgeable on Python packaging and distribution, but here's what I tried:
- Started with [src layout](https://www.maturin.rs/project_layout#alternate-python-source-directory-src-layout) that you linked before. I got confused about how to reconcile the 2 different `pyproject.toml`s that used different `build-backend`s.
- Found an example ([deltalake-python](https://github.com/delta-io/delta-rs/tree/main/python)) that had mixed Python and Rust. The organization looked clean, and used a single `maturin` `build-backend` for both Rust and Python files.
- Read through [maturin README](https://github.com/pyo3/maturin?tab=readme-ov-file#mixed-rustpython-projects) for more info as well. Main takeaway was that (1) the [deltalake-python](https://github.com/delta-io/delta-rs/tree/main/python) organization works and (2) it is good convention to prepend `_` to the Rust module if the intention is that it is internal.

That's how I ended up with the current structure.

I'm trying to fix the last `ci.yaml` `pyright` test, where it seems to say that `_lowtime_rust` does not exist. When I test locally, just running "pip install lowtime" seems to be sufficient in building both Rust and Python libraries, so I'm not sure why the automated test is failing. Will have to look deeper into this, but let me know if you spot anything.

**Comment by jaywonchung** [comment]
Have a `pyi` file? `lowtime/_lowtime_rs.pyi`.

```python
from __future__ import annotations

class PhillipsDessouky:
    node_ids: list[int]
    
    def __init__(
        self,
        node_ids: list[int],
        source_node_id: int,
        sink_node_id: int,
        edges_raw: list[tuple[tuple[int, int], float]]
    ) -> None: ...

    def max_flow(self) -> list[tuple[tuple[int, int], float]]: ...
```

**Comment by ohjuny** [comment]
> Have a `pyi` file? `lowtime/_lowtime_rs.pyi`.
> 
> ```python
> from __future__ import annotations
> 
> class PhillipsDessouky:
>     node_ids: list[int]
>     
>     def __init__(
>         self,
>         node_ids: list[int],
>         source_node_id: int,
>         sink_node_id: int,
>         edges_raw: list[tuple[tuple[int, int], float]]
>     ) -> None: ...
> 
>     def max_flow(self) -> list[tuple[tuple[int, int], float]]: ...
> ```

~DIdn't seem to work :( Will look further into this.~

Edit: needed to also add an empty `py.typed` file, documented in the [maturin docs](https://www.maturin.rs/project_layout#adding-python-type-information). Fun fact: there is an [open PR](https://github.com/PyO3/pyo3/issues/2454) on `pyo3` to generate these automatically.

Fixing some other `pyright` issues now.

File: .github/workflows/CI.yml
Comment (jaywonchung) [inline .github/workflows/CI.yml:1]: There is also a (lower case) `ci.yaml` in the directory. I think it's better if we can merge the two. Or maybe the lower case workflow is now stale.

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        with:
+          command: sdist
+          args: --out dist
+      - name: Upload sdist
+        uses: actions/upload-artifact@v4
+        with:
+          name: wheels-sdist
+          path: dist
+
+  release:
```
Comment (jaywonchung) [inline .github/workflows/CI.yml]: So, when we want to release a new version of lowtime with Rust bindings, will this workflow create a pre-compiled wheel for all these different kinds of platforms?

File: pyproject.toml [hunk 5]
```diff
@@ -17,6 +17,9 @@ classifiers = [
     "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
+    "Programming Language :: Rust",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Programming Language :: Python :: Implementation :: PyPy",
```
Comment (jaywonchung) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml
Comment (jaywonchung) [inline pyproject.toml:1]: Please add your name in the `project.authors` section.

File: pyproject.toml [hunk 6]
```diff
@@ -37,16 +44,6 @@ lint = ["ruff", "black"]
 test = ["pytest"]
 dev = ["ruff", "black", "pytest", "tyro", "pandas", "pyright"]
 
-[tool.setuptools]
-include-package-data = false
-
-[tool.setuptools.dynamic]
-version = {attr = "lowtime.__version__"}
```
Comment (jaywonchung) [inline pyproject.toml:44]: We used to dynamically parse the version of the project from the `lowtime.__version__` attribute inside `lowtime/__init__.py`. This part is also linked with line 30 (`dynamic = ["version"]`). It's fine if we give up dynamically parsing the library's version, but in any case the way we specify the version of the project should be consistent.

File: pyproject.toml.bak
Comment (jaywonchung) [inline pyproject.toml.bak:1]: Bro you should skim your PR changes at least once before requesting review

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+use pyo3::prelude::*;
+
+mod phillips_dessouky;
+use phillips_dessouky::PhillipsDessouky;
+
+#[pymodule]
+fn _lowtime_rs(m: &Bound<'_, PyModule>) -> PyResult<()> {
+    pyo3_log::init();  // send Rust logs to Python logger
```
Comment (jaywonchung) [inline src/lib.rs:8]: I think this is very convenient and being able to log messages using a unified Python logger is helpful for users as well. But I wonder what the performance implication of this is. If this requires some kind of Python-Rust context switch whenever logging messages need to be sent to Python, it might not be cheap.

File: src/phillips_dessouky.rs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+use pathfinding::directed::edmonds_karp::{
+    SparseCapacity,
+    Edge,
+    edmonds_karp
+};
+
+use std::time::{
+    Instant,
+    Duration,
+};
```
Comment (jaywonchung) [inline src/phillips_dessouky.rs]: Stuff like this doesn't have to be on multiple lines.
```suggestion
use std::time::{Instant, Duration};
```

File: src/phillips_dessouky.rs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        info!("PROFILING Rust_PhillipsDessouky::max_flow reformat OrderedFloat to f64 time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
+
+        Ok(flows_f64)
+    }
+}
+
+// Private to Rust, not exposed to Python
+impl PhillipsDessouky {
+    fn profile_duration(start: Instant, end: Instant) -> f64 {
+        let duration: Duration = end.duration_since(start);
```
Comment (jaywonchung) [inline src/phillips_dessouky.rs]:
```suggestion
        let duration = end.duration_since(start);
```
Return types are trivially inferred by the compiler. So adding these just add complexity to the code.

File: src/phillips_dessouky.rs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            &self.sink_node_id,
+            edges_edmonds_karp,
+        );
+        let profiling_end = Instant::now();
+        info!("PROFILING Rust_PhillipsDessouky::max_flow edmonds_karp time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
+
+        let profiling_start = Instant::now();
+        let flows_f64: Vec<((u32, u32), f64)> = flows.iter().map(|((from, to), flow)| {
+            ((*from, *to), flow.into_inner())
+        }).collect();
```
Comment (jaywonchung) [inline src/phillips_dessouky.rs]:
```suggestion
        let flows_f64: Vec<((u32, u32), f64)> = flows.into_iter().map(|(edge, flow)| {
            (edge, flow.into_inner())
        }).collect();
```
I think this copy can be avoided. Well, before you tell me that this doesn't compile ;)

File: src/phillips_dessouky.rs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        let profiling_end = Instant::now();
+        info!("PROFILING Rust_PhillipsDessouky::max_flow reformat OrderedFloat to f64 time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
```
Comment (jaywonchung) [inline src/phillips_dessouky.rs:62]: These `PROFILING` log messages are for dev purposes, and they need to be removed/prettified when we land it on `master`.

File: Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "_lowtime_rs"
+version = "0.1.0"
```
Comment (jaywonchung) [inline Cargo.toml]: Is it possible to not even have a version for this? If not, let's synchronize the version of this one with Python.

File: lowtime/.gitignore
Comment (jaywonchung) [inline lowtime/.gitignore:1]: `.gitignore` shouldn't be inside the Python package directory.

File: lowtime/_lowtime_rs.pyi [hunk 13]
```diff
@@ -0,0 +1,15 @@
+from __future__ import annotations
+
+import networkx as nx
+
+class PhillipsDessouky:
+    node_ids: list[int]
```
Comment (jaywonchung) [inline lowtime/_lowtime_rs.pyi]: Um what is this?

File: lowtime/_lowtime_rs.pyi [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+    def __init__(
+        self,
+        node_ids: list[int] | nx.classes.reportviews.NodeView,
+        source_node_id: int,
+        sink_node_id: int,
+        edges_raw: list[tuple[tuple[int, int], float]],
+    ) -> None: ...
+    def max_flow(self) -> list[tuple[tuple[int, int], float]]: ...
```
Comment (jaywonchung) [inline lowtime/_lowtime_rs.pyi:13]:
```suggestion
    ) -> None: ...
    
    def max_flow(self) -> list[tuple[tuple[int, int], float]]: ...
```

File: pyproject.toml.bak
Comment (jaywonchung) [inline pyproject.toml.bak:1]: Remove

**Review by jaywonchung — COMMENTED** [review]
This is absolutely great work, thank you! I've left some suggestions and comments.

File: Cargo.toml (hunk 12 above)
Comment (ohjuny) [reply Cargo.toml]: Seems to be removable. Not completely sure on the implications of not having a version, though.

File: lowtime/_lowtime_rs.pyi (hunk 13 above)
Comment (ohjuny) [reply lowtime/_lowtime_rs.pyi]: Do you mean the existence of this stray variable or something else?
For now, I've removed the stray variable here. I think this was a typo because the member variables are not exposed to Python anyway (the pyo3 getters have been removed Rust-side as well).
If you meant something else, let me know. Otherwise, feel free to mark this as resolved.

File: pyproject.toml
Comment (ohjuny) [reply pyproject.toml:1]: Yay :)

File: src/phillips_dessouky.rs (hunk 10 above)
Comment (jaywonchung) [reply src/phillips_dessouky.rs]: Updated a little bit. I don't think `edge` has to be unpacked into `(from, to)` as well.

File: pyproject.toml (hunk 6 above)
Comment (ohjuny) [reply pyproject.toml:44]: I am not quite sure what you mean (my knowledge on packaging/distribution is still kind of weak). I can see that `lowtime/__init__.py` specifies a version. I can see that we used to dynamically parse this version to set some (I'm guessing required) `setuptools` field. Since we change the `build-backend` to use `setuptools` instead, I'm not sure what `maturin` needs us to specify... Could you rephrase what you mean?

Comment (jaywonchung) [reply pyproject.toml:44]: Maturin currently does not support dynamic versions (https://github.com/PyO3/maturin/issues/199). Let's hardcode the version of lowtime to a specific number for now. When we release, we want to ensure that the version string of `lowtime.__version__`, Cargo.toml, and pyproject.toml are the same.

File: lowtime/_lowtime_rs.pyi [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+class PhillipsDessouky:
+    def __init__(
+        self,
+        node_ids: list[int] | nx.classes.reportviews.NodeView,
+        source_node_id: int,
+        sink_node_id: int,
+        edges_raw: list[tuple[tuple[int, int], float]],
+    ) -> None: ...
+
```
Comment (ohjuny) [inline lowtime/_lowtime_rs.pyi]: black formatting seems to not like the newline here :(

Comment (jaywonchung) [reply lowtime/_lowtime_rs.pyi]: Wow that's unexpected

File: src/lib.rs (hunk 7 above)
Comment (ohjuny) [reply src/lib.rs:8]: Yeah I was planning on keeping it for now along with the profiling logs, then possibly remove them once we don't need the profiling logs anymore (or profile the impact of logs and see if the impact is significant).

Comment (jaywonchung) [reply src/lib.rs:8]: Maybe remove all `info!` calls for one case and measure its e2e duration. If there's significant change, overhead might be high. I presume most of the execution time is in `edmonds_karp` anyway, though.

File: src/phillips_dessouky.rs (hunk 11 above)
Comment (ohjuny) [reply src/phillips_dessouky.rs:62]: Agreed. I will keep them until the end of this chain, then remove them. I think having a profiling infrastructure integrated into lowtime somehow could be a cool, separate project. I've made a note of this in the "Design Decisions" section of the main PR comment/overview as well.

Comment (jaywonchung) [reply src/phillips_dessouky.rs:62]: Maybe try running something like [Samply](https://github.com/mstange/samply/) on the program. For debug symbols I think you'll have to put `debug = true` under `[profile.release]` to get a flamegraph.

File: .github/workflows/CI.yml (hunk 4 above)
Comment (ohjuny) [reply .github/workflows/CI.yml]: This was auto-generated by maturin when I created the project with `maturin init`. I'm pretty sure it's doing what you mentioned (create a pre-compiled wheel the mentioned platforms), but not 100% certain. [Here](https://github.com/PyO3/maturin-action?tab=readme-ov-file) is some documentation/extra info.

Edit: feel free to mark as resolved if the `CI.yml` looks fine...

File: .github/workflows/CI.yml
Comment (ohjuny) [reply .github/workflows/CI.yml:1]: I think the lowercase`ci.yaml` isn't stale since it's doing the style/type checking (not done in the uppercase `CI.yml`. Since the `CI.yml` was autogenerated, I was hesitant to change it. But honestly I think it's fine to just merge the 2 files into one (renaming the `ci` job from `ci.yaml` to something like `style` on `CI.yml`). Wdyt?

Comment (jaywonchung) [reply .github/workflows/CI.yml:1]: Let's not touch the auto-generated one. We can rename the two files with more informative names.

File: src/phillips_dessouky.rs (hunk 11 above)
Comment (ohjuny) [reply src/phillips_dessouky.rs:62]: I haven't tried using Samply yet, but just based on running with/without Rust logger the profiling results look pretty much the same. In fact, for some runs, it seems to run slower without the Rust logger(?) I think this may just be inherent randomness involved in profiling large programs. and that we can assume there is no significant overhead.

Edit: feel free to mark as resolved if this sounds good.

File: .github/workflows/CI.yml
Comment (jaywonchung) [reply .github/workflows/CI.yml:1]: I'd rename this one to `maturin.yaml`.

**Review by jaywonchung — APPROVED** [review]
LGTM!

One thing that would be nice to have is an automated e2e test for result consistency.

File: .github/workflows/CI.yml (hunk 4 above)
Comment (jaywonchung) [reply .github/workflows/CI.yml]: I don't have the capacity now to read the maturin actions and generated YAML file. And in any case, I feel like we'll end up doing more than one releases to figure things out anyway. Let's leave it as is for now.
